File: 761.py
# https://aceptaelreto.com/problem/statement.php?id=761
# https://es.wikipedia.org/wiki/Algoritmo_de_Bellman-Ford


class Graph:
    def __init__(self, vertices):
        self.V = vertices  # Number of vertices
        self.edges = []  # List of edges

    def add_edge(self, u, v, w):
        self.edges.append((u, v, w))

    def remove_edges(self, u):
        # Removes all edges that start at u
        self.edges = [edge for edge in self.edges if edge[0] != u]

    def bellman_ford(self, src):
        # Step 1: Initialize distances from src to all other vertices as INFINITE
        dist = [float("inf")] * self.V
        dist[src] = 0

        # Step 2: Relax all edges |V| - 1 times
        for _ in range(self.V - 1):
            for u, v, w in self.edges:
                if dist[u] != float("inf") and dist[u] + w < dist[v]:
                    dist[v] = dist[u] + w

        # Step 3: Check for negative-weight cycles
        for u, v, w in self.edges:
            if dist[u] != float("inf") and dist[u] + w < dist[v]:
                return None

        return dist[self.V - 1]


import sys

data = sys.stdin.readlines()
index = 0

while index < len(data):
    columns, rows = [int(x) for x in data[index].strip().split()]
    index += 1

    matrix = []
    graph = Graph(columns * rows)
    wormholePositions = []

    # Store the map
    for i in range(rows):
        row = list(data[index].strip())
        matrix.append(row)
        index += 1

    # First, connect all valid adjacent cells
    for i in range(rows):
        for j in range(columns):
            if matrix[i][j] == "#":
                continue  # Wall, no connections
            u = i * columns + j
            # Check 4 directions
            for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
                ni, nj = i + dx, j + dy
                if 0 <= ni < rows and 0 <= nj < columns and matrix[ni][nj] != "#":
                    v = ni * columns + nj
                    graph.add_edge(u, v, 1)

            if matrix[i][j] == "O":
                wormholePositions.append(u)
                # Remove current outgoing edges (to simulate teleport)
                graph.remove_edges(u)

    for i in range(len(wormholePositions)):
        # col, row, value
        wormholeData = [int(x) for x in data[index].strip().split()]
        graph.add_edge(
            wormholePositions[i],
            (wormholeData[1] - 1) * columns + (wormholeData[0] - 1),
            wormholeData[2],
        )
        index += 1

    # Remove edges from the objective
    graph.remove_edges((rows - 1) * columns + (columns - 1))

    result = graph.bellman_ford(0)

    if result is None:
        print("EXPLOSION")
    elif result == float("inf"):
        print("IMPOSIBLE")
    else:
        print(result)

"""
Pista 1
Piensa cómo resolverías el problema si los agujeros de gusano nunca recargaran
la nave y por tanto no pudiera ocurrir el caso EXPLOSION.

Pista 2
El problema se puede modelizar como un grafo dirigido,
donde muchas aristas son implícitas, generadas por la vecindad de las celdas en el mapa.

Pista 3
Si los agujeros de gusano siempre consumieran energía,
el problema se podría resolver con el algoritmo de Dijkstra para
buscar el camino mínimo entre el origen y el destino.

Pista 4
El algoritmo de Dijkstra no funciona bien si hay aristas con pesos negativos.
Busca una alternativa.

Pista 5
El algoritmo de Bellman–Ford calcula el camino más corto desde un 
vértice de un grafo dirigido a todos los demás, 
y soporta la existencia de aristas con pesos negativos.
Permite además detectar la existencia de ciclos negativos
que hacen que no haya un camino más corto porque éste puede
siempre reducirse dando una vuelta más al ciclo.

Pista 6
Usa el algoritmo de Bellman-Ford para calcular el camino mínimo
y detectar si hay algún ciclo negativo. Si lo hay, escribe EXPLOSION.

Pista 7
Para que Jessy caiga en un ciclo negativo y haga explotar las 
baterías de su nave tiene que poder llegar a él. 
No siempre que hay un ciclo negativo hay que escribir EXPLOSION.
"""

File: 761.cpp
#include <iostream>
#include <vector>
#include <string>
#include <sstream>
#include <limits>
#include <tuple>

using namespace std;

class Graph
{
public:
    int V;
    vector<tuple<int, int, int>> edges;

    Graph(int vertices)
    {
        V = vertices;
    }

    void add_edge(int u, int v, int w)
    {
        edges.emplace_back(u, v, w);
    }

    void remove_edges(int u)
    {
        vector<tuple<int, int, int>> filtered;
        for (auto &edge : edges)
        {
            if (get<0>(edge) != u)
            {
                filtered.push_back(edge);
            }
        }
        edges = move(filtered);
    }

    pair<bool, int> bellman_ford(int src)
    {
        vector<int> dist(V, numeric_limits<int>::max());
        dist[src] = 0;

        for (int i = 0; i < V - 1; ++i)
        {
            for (const auto &edge : edges)
            {
                int u, v, w;
                tie(u, v, w) = edge;
                if (dist[u] != numeric_limits<int>::max() && dist[u] + w < dist[v])
                {
                    dist[v] = dist[u] + w;
                }
            }
        }

        for (const auto &edge : edges)
        {
            int u, v, w;
            tie(u, v, w) = edge;
            if (dist[u] != numeric_limits<int>::max() && dist[u] + w < dist[v])
            {
                return {false, 0}; // Negative cycle detected
            }
        }

        int target = V - 1;
        return {true, dist[target]};
    }
};

int main()
{
    string line;
    vector<string> data;
    while (getline(cin, line))
    {
        data.push_back(line);
    }

    size_t index = 0;

    while (index < data.size())
    {
        int columns, rows;
        {
            istringstream ss(data[index++]);
            ss >> columns >> rows;
        }

        vector<string> matrix;
        Graph graph(columns * rows);
        vector<int> wormholePositions;

        for (int i = 0; i < rows; ++i)
        {
            matrix.push_back(data[index++]);
        }

        for (int i = 0; i < rows; ++i)
        {
            for (int j = 0; j < columns; ++j)
            {
                if (matrix[i][j] == '#')
                    continue;

                int u = i * columns + j;

                // Connect to adjacent non-wall cells
                vector<pair<int, int>> directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
                for (size_t d = 0; d < directions.size(); ++d)
                {
                    int dx = directions[d].first;
                    int dy = directions[d].second;

                    int ni = i + dx;
                    int nj = j + dy;
                    if (ni >= 0 && ni < rows && nj >= 0 && nj < columns && matrix[ni][nj] != '#')
                    {
                        int v = ni * columns + nj;
                        graph.add_edge(u, v, 1);
                    }
                }

                if (matrix[i][j] == 'O')
                {
                    wormholePositions.push_back(u);
                    graph.remove_edges(u);
                }
            }
        }

        for (int i = 0; i < wormholePositions.size(); ++i)
        {
            int col, row, val;
            istringstream ss(data[index++]);
            ss >> col >> row >> val;
            int dest = (row - 1) * columns + (col - 1);
            graph.add_edge(wormholePositions[i], dest, val);
        }

        // Remove edges from the exit node
        graph.remove_edges((rows - 1) * columns + (columns - 1));

        auto result = graph.bellman_ford(0);
        if (!result.first)
        {
            cout << "EXPLOSION" << endl;
        }
        else if (result.second == numeric_limits<int>::max())
        {
            cout << "IMPOSIBLE" << endl;
        }
        else
        {
            cout << result.second << endl;
        }
    }

    return 0;
}
